### ISSUE TITLE
Need tooling to enforce style guide

### ISSUE THREAD

**Issue by SteveL-MSFT** [PowerShell/PowerShell#4708]
Everyone agrees that from a style perspective, the intent is to have consistency in the code to make it more readable and maintainable.   Update https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md as needed, but should be enforced by tooling and not by PR comments.

- [ ] enforcement on the CI side
- [x] enforcement on the developer side (part of Start-PSPester?)

**Comment by iSazonov** [comment]
We already have PR #3504 for CI side and request from @lzybkr for developer side (git commit hooks).

**Comment by SteveL-MSFT** [comment]
@iSazonov you can close this once that PR is merged.  Thanks!

**Comment by iSazonov** [comment]
The PR s problematic - we can't merge it until we fix all files in the repo.
Also I believe we should address git commit hooks in another PR - I never write such hooks.

/cc @lzybkr

**Comment by daxian-dbw** [comment]
@TravisEz13 Could you please comment? I think you have more experience on this and have some ideas about how we can gradually roll out the files that are covered by the formatting checks.

**Comment by iSazonov** [comment]
I see GitHub introduce `Checks`. Can we use this?

**Comment by TravisEz13** [comment]
I've seen other teams using https://www.codefactor.io/

**Comment by iSazonov** [comment]
Screenshots look ver interesting. Can we try this in the repo?

**Comment by TravisEz13** [comment]
@SteveL-MSFT @daxian-dbw @adityapatwardhan  What do you think?  Can we try codefactor.io?

**Comment by SteveL-MSFT** [comment]
@TravisEz13 I'm open to trying it out.  I'll set it up.

**Comment by adityapatwardhan** [comment]
I am open to trying it as long as it does not block PRs from being merged, at least to begin with.

**Comment by SteveL-MSFT** [comment]
https://www.codefactor.io/repository/github/powershell/powershell/overview/master

**Comment by iSazonov** [comment]
Thanks! Looks great.

> Have your own StyleCop rules? No problem, simply add Settings.StyleCop file to the root of this branch.

We can configure rules which we want.

**Comment by iSazonov** [comment]
[Overall grade](https://support.codefactor.io/i14-glossary) is C - Satisfactory. 
We could show the status on our main page. 
Perhaps we should work now to get into B status.

I selectively reviewed the results of `CodeFactor`. 
Most commonly found `Duplicate Code` and `Complex Method`. Others are rare.
Seems we can easily address most of CodeFactor Issues.

**Comment by iSazonov** [comment]
[Duplication](https://www.codefactor.io/repository/github/powershell/powershell/issues?category=Duplication&groupId=1) rules should be disabled. - it seems completely useless.

**Comment by iSazonov** [comment]
I believe we should disable unneeded rules and _fix_ most of style issues.
Motivation - conservative policy is very good for a closed project, for an open project we have to continuously discuss the style issues that slows code reviews and leads to the perplexity of contributors.

**Comment by markekraus** [comment]
@iSazonov I wouldn't call that rule useless. It shows code reuse where we could refactor to a utility method or something. But refactoring those issues may be high risk.

**Comment by markekraus** [comment]
@iSazonov I wasn't saying we shouldn't disable the rule, just that it wasn't completely useless. :)

**Comment by iSazonov** [comment]
We need new policy and roadmap.

**Comment by TravisEz13** [comment]
I agree it isn't useless, but many of the suggested refactoring to remove duplication is risky.  It would be nice to be able to leave it on as informational if that is an option.

**Comment by iSazonov** [comment]
> duplication 

I see only 519 issue so it's not intrusive and we can leave this as is.

**Comment by SteveL-MSFT** [comment]
@powercode We're open to trying other tools if you're aware of any.

**Comment by iSazonov** [comment]
CodeFactor is only control tool. We could use https://github.com/dotnet/codeformatter but it also requires a lot of effort.

**Comment by powercode** [comment]
I just think we should do this the other way around. Start with giving people the tools to format the source, and then start to enforce it on pull requests.

This is of course a matter of opinion, but I don't think the gain (if any) in readability and maintainability is worth the hassle.

I'm working on getting a ReSharper profile setup for the project that has formatting options that are at least close to Code Factor. I know it is a commercial tool, but a lot of C# devs has it. That could east the pain a bit. 

I installed StyleCop locally, but got lots of errors with spelling, and didn't figure out how to make it use the existing spelling dictionary.

In summary, it is not so much the tool I object to, as the backward order this is introduced in.
Tools first, then checks, not the other way around.

**Comment by iSazonov** [comment]
CodeFormatter doesn't work with PowerShell Core csproj files https://github.com/dotnet/codeformatter/issues/256

**Comment by iSazonov** [comment]
From @daxian-dbw 

> About the style rules, @TravisEz13, @adityapatwardhan and I went through all issues reported from CodeFactor a few weeks back, and we got a list of rules that we think should be turned off. Here is the list:
[1305: SA1305FieldNamesMustNotUseHungarianNotation](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1305.md)
[1502: SA1502ElementMustNotBeOnSingleLine](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1502.md)
[1310: SA1310FieldNamesMustNotContainUnderscore](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1310.md),
[1204: SA1204StaticElementsMustAppearBeforeInstanceElements](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1204.md)
[1009: SA1009ClosingParenthesisMustBeSpacedCorrectly](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1009.md)
[1501:SA1501StatementMustNotBeOnSingleLine](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1501.md)
[1513:SA1513ClosingBraceMustBeFollowedByBlankLine](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1513.md)
[1306:SA1306FieldNamesMustBeginWithLowerCaseLetter](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1306.md)
[1308:SA1308VariableNamesMustNotBePrefixed](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1308.md)
[1013:SA1013ClosingBracesMustBeSpacedCorrectly](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1013.md)
[1500:SA1500BracesForMultiLineStatementsMustNotShareLine](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1500.md)
[1010:SA1010OpeningSquareBracketsMustBeSpacedCorrectly](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1010.md)
[1026:SA1026CodeMustNotContainSpaceAfterNewKeywordInImplicitlyTypedArrayAllocation](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1026.md)
[1008:SA1008OpeningParenthesisMustBeSpacedCorrectly](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1008.md)
[1311:SA1311StaticReadonlyFieldsMustBeginWithUpperCaseLetter](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1311.md)
[1025:SA1025CodeMustNotContainMultipleWhitespaceInARow](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1025.md)
[1012:SA1012OpeningBracesMustBeSpacedCorrectly](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1012.md)
[1215:SA1215InstanceReadonlyElementsMustAppearBeforeInstanceNonReadonlyElements](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1215.md)
[1214:SA1214ReadonlyElementsMustAppearBeforeNonReadonlyElements](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1214.md)
[1210:SA1210UsingDirectivesMustBeOrderedAlphabeticallyByNamespace](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1210.md)
[1609:SA1609PropertyDocumentationMustHaveValue](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1609.md)

The rules about "one parameter per line" for method definition or invocation is not in this list, but in my opinion, we probably should disable that rule because 1) it doesn't bring much benefit -- as long as the parameters are well aligned, there is no readability issue; 2) there are too many instances in our existing code base that are violating that rule, and fixing all doesn't worth the time.

The following document rules make sense for public members, but they are applied to non-public members too, which causes a lot of noise data.

[1611:SA1611ElementParametersMustBeDocumented](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1611.md)
[1615:SA1615ElementReturnValueMustBeDocumented](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1615.md)
[1606:SA1606ElementDocumentationMustHaveSummaryText](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1606.md)

**Comment by iSazonov** [comment]
Sample of formatting with codeformatter that I compiled ##7263

**Comment by joeyaiello** [comment]
@SteveL-MSFT is this okay to close now?

**Comment by vexx32** [comment]
Do we have an omnisharp.json in the repo with the mentioned style rules set for VS Code users? Might be a good idea to do so if we haven't.

Also, I opened another issue already, but codefactor should definitely come back for those edge cases and misc recommendations.

**Comment by SteveL-MSFT** [comment]
@joeyaiello no, we need to integrate with CI

**Comment by iSazonov** [comment]
@vexx32 We have .editorconfig as unified config for all editors.
